storage/include/sparse_merkle/base_types.h

```diff
@@ -0,0 +1,129 @@
+// Concord
+//
+// Copyright (c) 2019 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").
+// You may not use this product except in compliance with the Apache 2.0 License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the sub-component's license, as noted in the
+// LICENSE file.
+
+#include <cstdint>
+#include <vector>
+
+#include "assertUtils.hpp"
+
+namespace concord {
+namespace storage {
+namespace sparse_merkle {
+
+// A nibble is 4 bits, stored in the lower bits of a byte.
+class Nibble {
+ public:
+  Nibble(uint8_t byte) {
+    Assert((byte & 0xF0) == 0);
+    data_ = byte;
+  }
+
+  bool get_bit(const size_t bit) const {
+    Assert(bit < 4);
+    return (data_ >> bit) & 1;
+  }
+
+  bool operator==(const Nibble& other) const { return data_ == other.data_; }
+
+  // Promote the underlying uint8_t to a size_t so that it can be used as an
+  // index into arrays and vectors.
+  size_t to_index() const { return data_; }
+
+  // Return the underlying representation
+  uint8_t data() const { return data_; }
+
+ private:
+  // Only the lower 4 bits are used
+  uint8_t data_;
+};
+
+// A Hash is a wrapper around a byte buffer containing a hash
+class Hash {
+ public:
+  // Change the following 2 constants if using a different hash algorithm
+  static const size_t SIZE_IN_BITS = 256;
+  const char* HASH_ALGORITHM = "SHA3-256";
```

Comment:
Can we either make it `static const char* const` or `static const std::string` so that the pointer is not present in every object?

Reply:
done.